src/json-patch-duplex.ts

```diff
@@ -524,6 +524,58 @@ module jsonpatch {
     return true;
   }
 
+  
+  /**
+   * Apply a single json-patch on an object tree
+   * Returns the result object.
+   */
+  export function applyPatch(tree: any, patch: any, validate = false): any {
+    if (validate && typeof tree !== 'object') {
+      throw new TypeError('Tree has to be an object');
+    }
+    // on root
+    if (patch.path === "") {
+      if (patch.op === 'add' || patch.op === 'replace') {// same for add or replace 
+        // a primitive value, just return it
+        if (typeof patch.value !== 'object') {
+          return patch.value;
+        } else if (typeof patch.value === 'object') {
+          // conflicting types [] vs {} just return the new value
+          if (_isArray(tree) !== _isArray(patch.value)) {
+            return patch.value;
+            //same type, use original apply
+          } else {
+            apply(tree, [patch], validate);
+            return tree;
+          }
+        }
+      } else if (patch.op === 'move' || patch.op === 'copy') { // it's a move or copy to root
+        // get the value by json-pointer in `from` field
+        var temp: any = { op: "_get", path: patch.from };
+        apply(tree, [temp]);
+        // to cover conflicting types and primitive values 
+        var newTree = applyPatch(tree, { op: 'replace', path: '', value: temp.value })
+        return newTree;
+      } else { // it's a remove or test on root
+        if (patch.op === 'test') {
+          return _equals(tree, patch.value);
+        } else { // a remove on root
```

Comment:
I would check `patch.op === 'remove'` here and and either error or ignore the patch (return the input) if the operation is anything else. Otherwise any invalid patch would result in a `null` result and it wouldn't be obvious why.